File: backend/core/text_splitter.py
"""Splits AI response text into streaming-friendly chunks."""

_SENTENCE_END = frozenset("。！？.!?")
_BREATH = frozenset("、,…")


class TextSplitter:
    def __init__(self, chunk_size: int = 50):
        self.chunk_size = chunk_size

    def split(self, text: str) -> list[str]:
        if not text:
            return []
        sentences = self._by_sentences(text)
        chunks: list[str] = []
        buf = ""
        for sentence in sentences:
            if len(sentence) > self.chunk_size:
                sub = self._by_breath(sentence)
                for s in sub:
                    if buf and len(buf + s) > self.chunk_size:
                        if buf.strip():
                            chunks.append(buf.strip())
                        buf = s
                    else:
                        buf += s
            else:
                if buf and len(buf + sentence) > self.chunk_size:
                    if buf.strip():
                        chunks.append(buf.strip())
                    buf = sentence
                else:
                    buf += sentence
        if buf.strip():
            chunks.append(buf.strip())
        return chunks

    def _by_sentences(self, text: str) -> list[str]:
        result, cur = [], ""
        for ch in text:
            cur += ch
            if ch in _SENTENCE_END:
                result.append(cur)
                cur = ""
        if cur:
            result.append(cur)
        return result

    def _by_breath(self, text: str) -> list[str]:
        result, cur = [], ""
        for ch in text:
            cur += ch
            if ch in _BREATH or len(cur) >= self.chunk_size:
                if cur.strip():
                    result.append(cur)
                cur = ""
        if cur:
            result.append(cur)
        return result

File: backend/src/core/text_splitter.rs
/// Splits AI-generated text into chunks suitable for incremental TTS
/// synthesis: prefer whole sentences, but break overly long sentences at
/// natural "breath" points (commas) so audio doesn't lag too far behind text.
const SENTENCE_ENDINGS: &[char] = &['。', '！', '？', '\n'];
const BREATH_POINTS: &[char] = &['、', ','];
const MAX_CHUNK_CHARS: usize = 60;

pub fn split(text: &str) -> Vec<String> {
    let sentences = split_keep_delimiters(text, SENTENCE_ENDINGS);

    sentences
        .into_iter()
        .flat_map(|sentence| {
            if sentence.chars().count() > MAX_CHUNK_CHARS {
                split_keep_delimiters(&sentence, BREATH_POINTS)
            } else {
                vec![sentence]
            }
        })
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect()
}

/// Splits `text` on any of `delimiters`, keeping the delimiter attached to
/// the preceding chunk.
fn split_keep_delimiters(text: &str, delimiters: &[char]) -> Vec<String> {
    let mut chunks = Vec::new();
    let mut current = String::new();

    for ch in text.chars() {
        current.push(ch);
        if delimiters.contains(&ch) {
            chunks.push(std::mem::take(&mut current));
        }
    }
    if !current.is_empty() {
        chunks.push(current);
    }

    chunks
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn splits_on_sentence_endings() {
        assert_eq!(split("こんにちは。元気ですか？"), vec!["こんにちは。", "元気ですか？"]);
    }

    #[test]
    fn splits_long_sentences_on_breath_points() {
        let long = "あ".repeat(70) + "、" + &"い".repeat(10) + "。";
        let result = split(&long);
        assert!(result.len() >= 2);
    }
}
